File: Leetcode/827.cpp
namespace DSU {
    vector<int> Size, father;
    void init(int n) {
        vector<int>().swap(Size);
        vector<int>().swap(father);
        Size.resize(n+5,1);
        father.resize(n+5,1);
        iota(father.begin(),father.end(),0);
    }
    int find(int x) {
        if (x != father[x]) {
            father[x] = find(father[x]);
        }
        return father[x];
    }
    void merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) {
            return;
        }
        if (Size[x] > Size[y]) {
            swap(x,y);
        }
        Size[y] += Size[x];
        father[x] = y;
    }
}

constexpr int STEPS = 4;
const int dx[] = {0,0,-1,1};
const int dy[] = {-1,1,0,0};

class Solution {
public:
    int largestIsland(vector<vector<int>>& grid) {
        int n = (int)grid.size();
        DSU::init(n*n);
        int new_i,new_j;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 1) {
                    for (int step = 0; step < STEPS; ++step) {
                        new_i = i+dx[step];
                        new_j = j+dy[step];
                        if (new_i >= 0 && new_i < n && new_j >= 0 && new_j < n && grid[new_i][new_j] == 1) {
                            DSU::merge(i*n+j,new_i*n+new_j);
                        }
                    }
                }
            }
        }
        int maxSize = 0;
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 1) {
                    maxSize = max(maxSize, DSU::Size[DSU::find(i*n+j)]);
                }
            }
        }
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 0) {
                    unordered_set<int> groupsAround;
                    for (int step = 0; step < STEPS; ++step) {
                        new_i = i+dx[step];
                        new_j = j+dy[step];
                        if (new_i >= 0 && new_i < n && new_j >= 0 && new_j < n && grid[new_i][new_j] == 1) {
                            groupsAround.insert(DSU::find(new_i*n+new_j));
                        }
                    } 
                    int temp = 1;   // add self
                    for (const auto &k : groupsAround) {
                        temp += DSU::Size[DSU::find(k)];
                    }
                    maxSize = max(maxSize, temp);
                }
            }
        }
        return maxSize;
    }
};

File: Leetcode/827.go
const STEPS int = 4
var (
	Size, father []int
	dx = [STEPS]int{0,0,-1,1}
	dy = [STEPS]int{-1,1,0,0}
)

func dsuInit(n int) {
	Size = nil
	father = nil
	Size = make([]int, n+5)
	father = make([]int, n+5)
	for i := 0; i < n+5; i++ {
		Size[i] = 1
	}
	for i := 0; i < n+5; i++ {
		father[i] = i
	}
}
func find(x int) int {
	if x != father[x] {
		father[x] = find(father[x])
	}
	return father[x]
}
func merge(x, y int){
	x = find(x)
	y = find(y)
	if x == y {
		return
	}
	if Size[x] > Size[y] {
		x, y = y, x
	}
	Size[y] += Size[x]
	father[x] = y
}
func largestIsland(grid [][]int) int {
    var n = len(grid)
    dsuInit(n*n)
    var newI, newJ int
    for i := 0; i < n; i++ {
        for j := 0; j < n; j++ {
            if grid[i][j] == 1 {
                for step := 0; step < STEPS; step++ {
                    newI = i+dx[step]
                    newJ = j+dy[step]
                    if newI >= 0 && newI < n && newJ >= 0 && newJ < n && grid[newI][newJ] == 1 {
                        merge(i*n+j,newI*n+newJ)
                    }
                }
            }
        }
    }
    var maxSize = 0
    var temp int
    for i := 0; i < n; i++ {
        for j := 0; j < n; j++ {
            if grid[i][j] == 1 {
                temp = Size[find(i*n+j)]
                if temp > maxSize {
                    maxSize = temp
                }
            }
        }
    }
    for i := 0; i < n; i++ {
        for j := 0; j < n; j++ {
            if grid[i][j] == 0 {
                temp = 1
                var unmap map[int]int
                unmap = make(map[int]int)
                for step := 0; step < STEPS; step++ {
                    newI = i+dx[step]
                    newJ = j+dy[step]
                    if newI >= 0 && newI < n && newJ >= 0 && newJ < n && grid[newI][newJ] == 1 {
                        unmap[find(newI*n+newJ)]++
                    }
                }
                for k, _ := range unmap {
                    temp += Size[find(k)]
                }
                if temp > maxSize {
                    maxSize = temp
                }
            }
        }
    }
    return maxSize
}